linera-core/src/local_node.rs

```diff
@@ -223,13 +223,48 @@ where
         }
 
         // Find the missing blobs locally and retry.
-        let required = match certificate.value() {
-            CertificateValue::ConfirmedBlock { executed_block, .. }
-            | CertificateValue::ValidatedBlock { executed_block, .. } => {
-                executed_block.required_blob_ids()
+        match certificate.value() {
+            CertificateValue::ConfirmedBlock { executed_block, .. } => {
+                self.check_missing_blob_ids(executed_block, missing_blob_ids)?;
+                let storage = self.storage_client();
+                Ok(storage
+                    .read_blobs(missing_blob_ids)
+                    .await?
+                    .into_iter()
+                    .flatten()
+                    .collect())
             }
-            CertificateValue::Timeout { .. } => HashSet::new(),
-        };
+            CertificateValue::ValidatedBlock { executed_block, .. } => {
+                let unique_missing_blob_ids =
+                    self.check_missing_blob_ids(executed_block, missing_blob_ids)?;
+                Ok(self
+                    .chain_state_view(chain_id)
+                    .await?
+                    .manager
+                    .get()
+                    .locked_blobs
+                    .iter()
+                    .filter(|(blob_id, _)| unique_missing_blob_ids.contains(blob_id))
+                    .map(|(_, blob)| blob.clone())
+                    .collect())
+            }
+            CertificateValue::Timeout { .. } => {
+                warn!(
+                    "validator requested blobs {:?} but they're not required",
+                    missing_blob_ids
+                );
+                Err(NodeError::InvalidChainInfoResponse)
+            }
+        }
+    }
```

Comment:
This method needs a comprehensive test coverage with each test case clearly showing _what_ we're trying to achieve. I don't think the "requirements"/rules for these are written down anywhere? i.e. It should be clear to the team (and based on the code) when a blob is expected to be in one of the collections (locked_blobs, proposed_blobs, storage).